File: src/main/java/org/polyfrost/glintcolorizer/util/GlintMetadata.java
package org.polyfrost.glintcolorizer.util;

import org.polyfrost.glintcolorizer.config.GlintColorizerConfig;
import org.polyfrost.glintcolorizer.config.category.BaseGlint;
import org.polyfrost.glintcolorizer.config.category.ShinyPotsCategory;
import net.minecraft.core.component.DataComponents;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.item.PotionItem;

import java.util.Objects;

public class GlintMetadata {
	public enum RenderMode {
		HELD,
		SHINY,
		GUI,
		DROPPED,
		FRAMED
	}

	private static RenderMode cachedRenderMode = RenderMode.HELD;
	private static ItemStack cachedItemStack = ItemStack.EMPTY;

	public static void setRenderMode(RenderMode renderMode) {
		if (!(renderMode == RenderMode.GUI && cachedRenderMode == RenderMode.SHINY)) {
			// preserve shiny render mode !
			cachedRenderMode = renderMode;
		}
	}

	public static RenderMode getRenderMode() {
		return cachedRenderMode;
	}

	public static void setItemStack(ItemStack itemStack) {
		if (!ItemStack.matches(itemStack, cachedItemStack)) {
			cachedItemStack = itemStack;
		}
	}

	public static ItemStack getItemStack() {
		return cachedItemStack;
	}

	public static BaseGlint getRenderingOptions() {
		return switch (cachedRenderMode) {
			case HELD -> GlintColorizerConfig.heldItemGlint;
			case SHINY -> GlintColorizerConfig.shinyPots;
			case GUI -> GlintColorizerConfig.guiItemGlint;
			case DROPPED -> GlintColorizerConfig.droppedItemGlint;
			case FRAMED -> GlintColorizerConfig.framedItemGlint;
		};
	}

	public static int getGlintColor(GlintLayer layer, boolean isArmor) {
		BaseGlint options = isArmor ? GlintColorizerConfig.armorGlint : getRenderingOptions();
		if (cachedItemStack.getItem() instanceof PotionItem && GlintColorizerConfig.shinyPots.useCustomColor) {
			options = GlintColorizerConfig.shinyPots;
			if (options instanceof ShinyPotsCategory shinyPotsCategory && shinyPotsCategory.usePotionBasedColor && cachedItemStack.has(DataComponents.POTION_CONTENTS)) {
				return Objects.requireNonNull(cachedItemStack.getComponents().get(DataComponents.POTION_CONTENTS)).getColor();
			}
		}

		return (options.individualStrokes
				? (layer == GlintLayer.FIRST ? options.strokeOneColor : options.strokeTwoColor)
				: options.color).getArgb();
	}
}

File: src/main/kotlin/org/polyfrost/glintcolorizer/GlintMetadata.kt
package org.polyfrost.glintcolorizer

import net.minecraft.item.ItemPotion
import net.minecraft.item.ItemStack
import org.polyfrost.glintcolorizer.config.BaseGlint
import org.polyfrost.glintcolorizer.config.GlintConfig

object GlintMetadata {
    private val EMPTY =
        //#if MC <=1.12.2
        ItemStack(net.minecraft.init.Blocks.air)
        //#else
        //$$ItemStack.EMPTY
        //#endif

    enum class RenderMode {
        HELD,
        SHINY,
        GUI,
        DROPPED,
        FRAMED
    }

    private var renderingItemCached: ItemStack = EMPTY

    @JvmStatic
    var renderMode = RenderMode.HELD
        set(value) {
            if (!(value == RenderMode.GUI && field == RenderMode.SHINY)) {
                field = value // preserve shiny render mode !
            }
        }

    @JvmStatic
    fun setItemStack(itemStack: ItemStack?) {
        var tempStack = itemStack
        if (tempStack == null) {
            tempStack = EMPTY
        }

        if (!ItemStack.areItemStacksEqual(tempStack, renderingItemCached)) {
            renderingItemCached = tempStack
        }
    }

    @JvmStatic
    fun getRenderingOptions(): BaseGlint {
        return when (renderMode) {
            RenderMode.HELD -> GlintConfig.heldItemOptions
            RenderMode.SHINY -> GlintConfig.shinyPotsOptions
            RenderMode.GUI -> GlintConfig.guiItemOptions
            RenderMode.DROPPED -> GlintConfig.droppedItemOptions
            RenderMode.FRAMED -> GlintConfig.framedItemOptions
        }
    }

    @JvmStatic
    fun getGlintColor(layer: GlintLayer, isArmor: Boolean): Int {
        var options = if (isArmor) GlintConfig.armorOptions else getRenderingOptions()
        if (renderingItemCached.item is ItemPotion && GlintConfig.shinyPotsOptions.useCustomColor) {
            options = GlintConfig.shinyPotsOptions
            if (options.usePotionBasedColor) {
                //#if MC <=1.21.5
                //$$
                //#if MC <= 1.12.2
                //$$return net.minecraft.potion.PotionUtil.getColor(renderingItemCached!!) or -0x1000000
                //#elseif MC <=1.8.9
                return net.minecraft.potion.PotionHelper.getLiquidColor(
                    renderingItemCached.metadata,
                    false
                ) or -0x1000000
                //#endif
            }
        }

        if (options.individualStrokes) {
            if (layer == GlintLayer.FIRST) {
                return options.strokeOneColor.rgba;
            } else if (layer == GlintLayer.SECOND) {
                return options.strokeTwoColor.rgba;
            }
        }

        return options.glintColor.rgba
    }
}
